File: advent-of-code-2017/02_2.clj
(require '[clojure.string :as str])

(defn result [xs]
  (let [[x y] (reduce (fn [acc x]
                        (reduce (fn [yacc y]
                                  (cond (not (empty? yacc)) yacc
                                        (and (not (= x y)) (= 0 (mod x y))) [x y]
                                        :else []))
                                acc
                                xs))
                      []
                      xs)]
    (/ x y)))

(defn parse-row [r]
  (map read-string (str/split r #"\t")))

(defn solve []
  (->> (slurp "02_input.tsv")
       (#(str/split % #"\n"))
       (map (comp result parse-row))
       (reduce +)))

(println (solve))

File: advent-of-code-2017/02_2.hs
import Control.Arrow ((&&&))
import Data.List.Split (splitOn)

divisible :: [Int] -> Int -> [Int] -> [Int]
divisible xs n acc =
  foldr
    (\x acc' ->
      if (not (null acc'))
        then acc'
        else
          if (n /= x && n `mod` x == 0)
            then [n, x]
            else [])
    acc
    xs

result :: [Int] -> Int
result xs =
  let [x, y] = foldr (divisible xs) [] xs
  in x `div` y

parseRow :: String -> [Int]
parseRow = map read . splitOn "\t"

main = do
  print . sum . map (result . parseRow) . lines =<< readFile "02_input.tsv"
